Add multiplayer beacons

ZivDero opened the pull request:
## Beacons

Vinifera implements multiplayer beacons, similar to those seen in Red Alert 2.

In `RULES.INI`:
```ini
[General]
BeaconsEnabled=no  ; boolean, are beacons enabled?
SPBeacons=no       ; boolean, can the player place beacons in single-player?
MaxBeacons=-1      ; integer, maximum beacons per player. When the cap is reached, the oldest beacon will be deleted when a new beacon is placed. Negative numbers mean there is no cap.
```

- When a beacon is placed/detected, a sound effect and an EVA line are played.

```ini
[AudioVisual]
PlaceBeaconSound=   ; VocType, the sound played when the player places a beacon.
PlaceBeaconVoice=   ; VoxType, the EVA line played when the player places a beacon.
DetectBeaconVoice=  ; VoxType, the EVA line played another player places a beacon.
```

- Beacons can have a text label.
- Additionally, beacons may be placed with a preset label by holding `CTRL`, `SHIFT`, `ALT`, or a combination of them. The text that is shows as the preview, and the text that is set on the placed beacon can be configured.

In `UI.INI`:
```ini
[Ingame]
BeaconTextOffset=32         ; integer, the Y offset for the beacon text.
BeaconPreviewTextOffset=20  ; integer, the Y offset for the beacon text during preview.

BeaconText1=Expand  ; string, shown when placing with Shift.
BeaconText2=Attack  ; string, shown when placing with Ctrl.
BeaconText3=Move    ; string, shown when placing with Alt.
BeaconText4=        ; string, shown when placing with Ctrl+Shift.
BeaconText5=        ; string, shown when placing with Alt+Shift.
BeaconText6=Defend  ; string, shown when placing with Ctrl+Alt.
BeaconText7=        ; string, shown when placing with Ctrl+Alt+Shift.

BeaconPreviewText1=Expand  ; string, preview shown with Shift.
BeaconPreviewText2=Attack  ; string, preview shown with Ctrl.
BeaconPreviewText3=Move    ; string, preview shown with Alt.
BeaconPreviewText4=        ; string, preview shown with Ctrl+Shift.
BeaconPreviewText5=        ; string, preview shown with Alt+Shift.
BeaconPreviewText6=Defend  ; string, preview shown with Ctrl+Alt.
BeaconPreviewText7=        ; string, preview shown with Ctrl+Alt+Shift.
```

- When drawn on the map, beacons use `PBEACON.SHP`. The animation is drawn using the unit palette, and is remapped to the beacon owner's house color. The second half of the animation is played when the beacon is selected.

- When drawn on the radar, beacons use `RDRBEACN.SHP`.

- Beacon animations are not tied to game FPS and instead play at a set framerate.

In `UI.INI`:
```ini
[Ingame]
BeaconAnimFramesPerSecond=25       ; integer, the framerate at which the beacon anim is played.
RadarBeaconAnimFramesPerSecond=25  ; integer, the framerate at which the radar beacon anim is played.
```

- Beacon placement/selection uses new mouse actions. For their defaults, please refer to Actions.

#NOTE:
RA2 graphics work with this PR.
[shapes.zip](https://github.com/user-attachments/files/22875281/shapes.zip)

Discussion #1475 

The following game.exe should be used with this PR:
[game.zip](https://github.com/user-attachments/files/22888639/game.zip)

ZivDero committed:
123

ZivDero committed:
Untie the animations from the framerate

ZivDero committed:
Fix MP, add beacon text presets

ZivDero committed:
Cleanup

ZivDero committed:
Docs

ZivDero committed:
Remove duplicate in docs

ZivDero committed:
Add some missed comments

Metadorius commented on src/extensions/conquerext/conquerext_hooks.cpp:
```diff
@@ -0,0 +1,485 @@
+    **  4 = caller should send the Overflow buffer
+    */
+
+    if (rc == 1 || rc == 2) {
```
this begs to be an enum

Metadorius commented on src/extensions/scenario/scenarioext.cpp:
```diff
@@ -247,6 +248,8 @@ bool ScenarioClassExtension::Read_INI(CCINIClass &ini)
      */
     Read_Tutorial_INI(ini, true);
 
+    BeaconManager.Load_Art();
```
in YR, loading a saved game will cause becom placement to crash the game due to no call to Load Art when loading the game, have you already accounted for that?

ZivDero replied:
Yeah

Metadorius commented on src/extensions/tactical/tacticalext.cpp:
```diff
@@ -556,14 +562,24 @@ void TacticalExtension::Render_Post()
     Draw_Super_Timers();
+
+    /**
+     *  In decon placement mode, holding modifier keys can give you a preset text
```
typo

ZivDero replied:
Fixed

Metadorius reviewed (commented):
hard to review when semi-related refactorings are also present

ZivDero committed:
Fix typo

ZivDero committed:
Merge branch 'develop' into feature/beacons

Metadorius commented on docs/User-Interface.md:
````diff
@@ -77,6 +77,48 @@ Due to limitations of the game's tooltip system, the length of the description i
+MaxBeacons=-1        ; integer, maximum beacons per player. When the cap is reached, the oldest beacon will be deleted when a new beacon is placed. Negative numbers mean there is no cap.
+```
+
+- When a beacon is placed/detected, a sound effect an EVA line are played.
````
> a sound effect an EVA line are played
typo

ZivDero committed:
Fix typos

ZivDero committed:
Enable beacons in skirmish

ZivDero committed:
Misc fixes

ZivDero committed:
Fix filename in docs

JoyfulShush commented:
All in all, works nicely. There a few issues to handle, mostly cosmetic. As those issues usually include a screenshot or a video, I'll post links to Discord instead of bringing them here.

* If text message is too long, beacon's message will be cleared out and instead would begin from scratch, removing everything that you typed so far (should instead limit you from writing more)
* When you scroll down on beacons the text of the beacons seems to gather on top. See [screenshot](https://discord.com/channels/338055131666972683/826473223276986409/1427304595649986692)
* Score screen crash - most likely not related to the feature, but worth validating. Only occurred once, in all other instances works as expected. See [debug files](https://discord.com/channels/338055131666972683/826473223276986409/1427308179691409410).
* If text message is too long, beacon's message will be cleared out and instead would begin from scratch, removing everything that you typed so fa. Should instead limit you from writing more. Check if this is something that can be done.
* If you delete another player's beacon, the game still changes your cursor whenever you hover where the beacon was, until you exit the cursor. Doesn't happen on your own beacons. See [video](https://discord.com/channels/338055131666972683/826473223276986409/1427313581686394990)
* Other players in your team can delete your beacons. According to @ZivDero , this might be a problem due to the way the game handles beacon. To verify and check.
* As spectator, the cursor changes to beacon place/select cursor even though the spectator cannot see the beacons. Can't interact or delete the beacon. See [screenshot](https://discord.com/channels/338055131666972683/826473223276986409/1427316919815372882)
* When you have radar, as you type each letter produces a radar event. See [video](https://discord.com/channels/338055131666972683/826473223276986409/1427320451771928637)
* While a beacon is selected, pressing F8 will show "To All:" and let you type but it actually doesn't send a message to anyone, instead just writes it on the beacon. See [video](https://discord.com/channels/338055131666972683/826473223276986409/1427358026368811249)

Minor enhancements:
* Introduce a cooldown to the EVA messages for beacon pings. Teammates can be annoying by placing a ton of beacons, spamming the EVA sounds.
* After pressing Enter to set a message, deselect the beacon.

ZivDero committed:
Various fixes and improvements

ZivDero commented:
> When you scroll down on beacons the text of the beacons seems to gather on top. See [screenshot](https://discord.com/channels/338055131666972683/826473223276986409/1427304595649986692)

Fixed.

> Score screen crash - most likely not related to the feature, but worth validating. Only occurred once, in all other instances works as expected. See [debug files](https://discord.com/channels/338055131666972683/826473223276986409/1427308179691409410).

Unrelated indeed, the same thing @Rampastring 's been battling for a while.

> If you delete another player's beacon, the game still changes your cursor whenever you hover where the beacon was, until you exit the cursor. Doesn't happen on your own beacons. See [video](https://discord.com/channels/338055131666972683/826473223276986409/1427313581686394990)

Shouldn't be able to get the beacon cursor over beacons you can't see anymore.

> Other players in your team can delete your beacons. According to @ZivDero , this might be a problem due to the way the game handles beacon. To verify and check.

You should still be able to delete other players' beacons, but no long be able to mouse over them.

> As spectator, the cursor changes to beacon place/select cursor even though the spectator cannot see the beacons. Can't interact or delete the beacon. See [screenshot](https://discord.com/channels/338055131666972683/826473223276986409/1427316919815372882)

Should be fixed.

> When you have radar, as you type each letter produces a radar event. See [video](https://discord.com/channels/338055131666972683/826473223276986409/1427320451771928637)

Fixed.

> While a beacon is selected, pressing F8 will show "To All:" and let you type but it actually doesn't send a message to anyone, instead just writes it on the beacon. See [video](https://discord.com/channels/338055131666972683/826473223276986409/1427358026368811249)

Should be fixed. Additionally, `Enter` now also works for "To All:" and `backspace` works for "To Allies:".

> After pressing Enter to set a message, deselect the beacon.

Sure, should be done.

ZivDero committed:
Adjust message formats

ZivDero committed:
Allies -> Team

ZivDero committed:
Remove outdated enum

ZivDero committed:
Use references

ZivDero committed:
Store IDs in a beacon

ZivDero committed:
Cleanup

ZivDero committed:
Add customizable beacon text

ZivDero committed:
Update message prefixes

JoyfulShush commented:
> When you scroll down on beacons the text of the beacons seems to gather on top. See [screenshot](https://discord.com/channels/338055131666972683/826473223276986409/1427304595649986692)

> Fixed.

Confirmed.

> If you delete another player's beacon, the game still changes your cursor whenever you hover where the beacon was, until you exit the cursor. Doesn't happen on your own beacons. See [video](https://discord.com/channels/338055131666972683/826473223276986409/1427313581686394990)

> Shouldn't be able to get the beacon cursor over beacons you can't see anymore.

Confirmed.

> Other players in your team can delete your beacons. According to @ZivDero , this might be a problem due to the way the game handles beacon. To verify and check.

> You should still be able to delete other players' beacons, but no long be able to mouse over them.

Confirmed.

> As spectator, the cursor changes to beacon place/select cursor even though the spectator cannot see the beacons. Can't interact or delete the beacon. See [screenshot](https://discord.com/channels/338055131666972683/826473223276986409/1427316919815372882)

> Should be fixed.

Confirmed.

> When you have radar, as you type each letter produces a radar event. See [video](https://discord.com/channels/338055131666972683/826473223276986409/1427320451771928637)

> Fixed.

Confirmed.

> While a beacon is selected, pressing F8 will show "To All:" and let you type but it actually doesn't send a message to anyone, instead just writes it on the beacon. See [video](https://discord.com/channels/338055131666972683/826473223276986409/1427358026368811249)

> Should be fixed. Additionally, Enter now also works for "To All:" and backspace works for "To Allies:".

Confirmed.

> After pressing Enter to set a message, deselect the beacon.

> Sure, should be done.

Confirmed.

ZivDero committed:
Docs fix

ZivDero committed:
Fix bugs with beacon text

ZivDero committed:
Fix deleting beacons when the player dies.

ZivDero committed:
Disable message overflow

JoyfulShush commented:
All issues were resolved, and testing has been concluded with high success. 

A couple bugs were found in the last testing round:

**Bug:** Select your own beacon, start typing, don't send the message. Press right click to deselect the beacon. Then continue typing and send.
this causes two issues:
* It sends the whole message as a chat message. It saves whatever you had at the time you deselected the beacon as the text.
* It causes phantom cursors to show up in random positions on the map. We found cursors at map edges, or near our bases. Not sure where it decides the cursor position.

**Expected:** Deselecting the beacon exits the message and sets it as the beacon's message. No phantom cursors should show up in the map.

**Bug:** You can click on another player's beacon, press enter while the beacon's selected, type something, then press enter. 

* The beacon's text will change accordingly for you, but not to any other player.
* You'll also send the beacon's text to all.

**Expected:** - you should not be able to change other's beacon text, even while selecting someone else's beacon.

ZivDero committed:
Rework how input is handled

ZivDero commented:
> All issues were resolved, and testing has been concluded with high success.
> 
> A couple bugs were found in the last testing round:
> 
> **Bug:** Select your own beacon, start typing, don't send the message. Press right click to deselect the beacon. Then continue typing and send. this causes two issues:
> 
> * It sends the whole message as a chat message. It saves whatever you had at the time you deselected the beacon as the text.
> * It causes phantom cursors to show up in random positions on the map. We found cursors at map edges, or near our bases. Not sure where it decides the cursor position.
> 
> **Expected:** Deselecting the beacon exits the message and sets it as the beacon's message. No phantom cursors should show up in the map.
> 
> **Bug:** You can click on another player's beacon, press enter while the beacon's selected, type something, then press enter.
> 
> * The beacon's text will change accordingly for you, but not to any other player.
> * You'll also send the beacon's text to all.
> 
> **Expected:** - you should not be able to change other's beacon text, even while selecting someone else's beacon.

1) Pressing RMB will now work the same as pressing ESC, that is, will it abort typing and clear the beacon's text.
2) You can no longer change the text for another player's beacon, pressing ENTER will instead start a normal chat message

ZivDero committed:
Refactor FontClass, textprint.h

ZivDero committed:
Merge branch 'develop' into feature/beacons

ZivDero committed:
Adjust for TS++ changes

ZivDero committed:
Merge branch 'develop' into feature/beacons

ZivDero committed:
Merge branch 'develop' into feature/beacons

ZivDero committed:
Post-merge fix

ZivDero committed:
Merge remote-tracking branch 'origin/develop' into feature/beacons
